File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/STT/RunAnywhere+STT.swift
//
//  RunAnywhere+STT.swift
//  RunAnywhere SDK
//
//  Public API for Speech-to-Text operations.
//  All transcription flows through C++ via CppBridge.STT / rac_stt_component,
//  which provides automatic telemetry for every backend (ONNX, WhisperKit, etc.).
//

@preconcurrency import AVFoundation
import CRACommons
import Foundation
import os

// MARK: - STT Operations

public extension RunAnywhere {

    // MARK: - Simple Transcription

    /// Simple voice transcription using default model
    /// - Parameter audioData: Audio data to transcribe
    /// - Returns: Transcribed text
    static func transcribe(_ audioData: Data) async throws -> String {
        guard isInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }
        try await ensureServicesReady()

        let result = try await transcribeWithOptions(audioData, options: STTOptions())
        return result.text
    }

    // MARK: - Model Loading

    /// Unload the currently loaded STT model
    static func unloadSTTModel() async throws {
        guard isSDKInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        await CppBridge.STT.shared.unload()
    }

    /// Check if an STT model is loaded
    static var isSTTModelLoaded: Bool {
        get async {
            return await CppBridge.STT.shared.isLoaded
        }
    }

    // MARK: - Transcription

    /// Transcribe audio data to text (with options)
    /// - Parameters:
    ///   - audioData: Raw audio data
    ///   - options: Transcription options
    /// - Returns: Transcription output with text and metadata
    static func transcribeWithOptions(
        _ audioData: Data,
        options: STTOptions
    ) async throws -> STTOutput {
        guard isSDKInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        let handle = try await CppBridge.STT.shared.getHandle()

        guard await CppBridge.STT.shared.isLoaded else {
            throw SDKError.stt(.notInitialized, "STT model not loaded")
        }

        let modelId = await CppBridge.STT.shared.currentModelId ?? "unknown"
        let startTime = Date()

        let audioSizeBytes = audioData.count
        let audioLengthSec = estimateAudioLength(dataSize: audioSizeBytes)

        var sttResult = rac_stt_result_t()
        defer { rac_stt_result_free(&sttResult) }
        let transcribeResult = options.withCOptions { cOptionsPtr in
            audioData.withUnsafeBytes { audioPtr in
                rac_stt_component_transcribe(
                    handle,
                    audioPtr.baseAddress,
                    audioData.count,
                    cOptionsPtr,
                    &sttResult
                )
            }
        }

        guard transcribeResult == RAC_SUCCESS else {
            throw SDKError.stt(.processingFailed, "Transcription failed: \(transcribeResult)")
        }

        let endTime = Date()
        let processingTimeSec = endTime.timeIntervalSince(startTime)

        let transcribedText: String
        if let textPtr = sttResult.text {
            transcribedText = String(cString: textPtr)
        } else {
            transcribedText = ""
        }
        let detectedLanguage: String?
        if let langPtr = sttResult.detected_language {
            detectedLanguage = String(cString: langPtr)
        } else {
            detectedLanguage = nil
        }
        let confidence = sttResult.confidence

        let metadata = TranscriptionMetadata(
            modelId: modelId,
            processingTime: processingTimeSec,
            audioLength: audioLengthSec
        )

        return STTOutput(
            text: transcribedText,
            confidence: confidence,
            wordTimestamps: nil,
            detectedLanguage: detectedLanguage,
            alternatives: nil,
            metadata: metadata
        )
    }

    /// Transcribe audio buffer to text
    /// - Parameters:
    ///   - buffer: Audio buffer
    ///   - language: Optional language hint
    /// - Returns: Transcription output
    static func transcribeBuffer(
        _ buffer: AVAudioPCMBuffer,
        language: String? = nil
    ) async throws -> STTOutput {
        guard isSDKInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        guard let channelData = buffer.floatChannelData else {
            throw SDKError.stt(.emptyAudioBuffer, "Audio buffer has no channel data")
        }

        let frameLength = Int(buffer.frameLength)
        let audioData = Data(bytes: channelData[0], count: frameLength * MemoryLayout<Float>.size)

        let options: STTOptions
        if let language = language {
            options = STTOptions(language: language)
        } else {
            options = STTOptions()
        }

        return try await transcribeWithOptions(audioData, options: options)
    }

    /// Start streaming transcription
    @available(*, deprecated, message: "Use transcribeStream(audioData:options:onPartialResult:) instead")
    static func startStreamingTranscription(
        options _: STTOptions = STTOptions(),
        onPartialResult _: @escaping (STTTranscriptionResult) -> Void,
        onFinalResult _: @escaping (STTOutput) -> Void,
        onError _: @escaping (Error) -> Void
    ) async throws {
        throw SDKError.stt(.streamingNotSupported, "Use transcribeStream(audioData:options:onPartialResult:) instead")
    }

    /// Transcribe audio with streaming callbacks
    static func transcribeStream(
        audioData: Data,
        options: STTOptions = STTOptions(),
        onPartialResult: @escaping (STTTranscriptionResult) -> Void
    ) async throws -> STTOutput {
        guard isSDKInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        let handle = try await CppBridge.STT.shared.getHandle()

        guard await CppBridge.STT.shared.isLoaded else {
            throw SDKError.stt(.notInitialized, "STT model not loaded")
        }

        guard await CppBridge.STT.shared.supportsStreaming else {
            throw SDKError.stt(.streamingNotSupported, "Model does not support streaming")
        }

        let modelId = await CppBridge.STT.shared.currentModelId ?? "unknown"
        let startTime = Date()

        let context = STTStreamingContext(onPartialResult: onPartialResult)
        let contextPtr = Unmanaged.passRetained(context).toOpaque()

        let result = options.withCOptions { cOptionsPtr in
            audioData.withUnsafeBytes { audioPtr in
                rac_stt_component_transcribe_stream(
                    handle,
                    audioPtr.baseAddress,
                    audioData.count,
                    cOptionsPtr,
                    { partialText, isFinal, userData in
                        guard let userData = userData else { return }
                        let ctx = Unmanaged<STTStreamingContext>.fromOpaque(userData).takeUnretainedValue()

                        let text = partialText.map { String(cString: $0) } ?? ""
                        let partialResult = STTTranscriptionResult(
                            transcript: text,
                            confidence: nil,
                            timestamps: nil,
                            language: nil,
                            alternatives: nil
                        )

                        ctx.onPartialResult(partialResult)

                        if isFinal == RAC_TRUE {
                            ctx.finalText = text
                        }
                    },
                    contextPtr
                )
            }
        }

        let finalContext = Unmanaged<STTStreamingContext>.fromOpaque(contextPtr).takeRetainedValue()

        guard result == RAC_SUCCESS else {
            throw SDKError.stt(.processingFailed, "Streaming transcription failed: \(result)")
        }

        let endTime = Date()
        let processingTimeSec = endTime.timeIntervalSince(startTime)
        let audioLengthSec = estimateAudioLength(dataSize: audioData.count)

        let metadata = TranscriptionMetadata(
            modelId: modelId,
            processingTime: processingTimeSec,
            audioLength: audioLengthSec
        )

        return STTOutput(
            text: finalContext.finalText,
            confidence: 0.0,
            wordTimestamps: nil,
            detectedLanguage: nil,
            alternatives: nil,
            metadata: metadata
        )
    }

    /// Process audio samples for streaming transcription
    static func processStreamingAudio(_ samples: [Float], options: STTOptions = STTOptions()) async throws {
        guard isSDKInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        let handle = try await CppBridge.STT.shared.getHandle()

        guard await CppBridge.STT.shared.isLoaded else {
            throw SDKError.stt(.notInitialized, "STT model not loaded")
        }

        let data = samples.withUnsafeBufferPointer { Data(buffer: $0) }

        var sttResult = rac_stt_result_t()
        defer { rac_stt_result_free(&sttResult) }
        let transcribeResult = options.withCOptions { cOptionsPtr in
            data.withUnsafeBytes { audioPtr in
                rac_stt_component_transcribe(
                    handle,
                    audioPtr.baseAddress,
                    data.count,
                    cOptionsPtr,
                    &sttResult
                )
            }
        }

        if transcribeResult != RAC_SUCCESS {
            throw SDKError.stt(.processingFailed, "Streaming process failed: \(transcribeResult)")
        }
    }

    /// Stop streaming transcription
    static func stopStreamingTranscription() async {
        // No-op - streaming is handled per-call
    }

    // MARK: - Private Helpers

    /// Estimate audio length from data size (assumes 16kHz mono 16-bit)
    private static func estimateAudioLength(dataSize: Int) -> Double {
        let bytesPerSample = 2  // 16-bit
        let sampleRate = 16000.0
        let samples = Double(dataSize) / Double(bytesPerSample)
        return samples / sampleRate
    }
}

// MARK: - Streaming Context Helper

/// Context class for bridging C callbacks to Swift closures.
private final class STTStreamingContext: Sendable {
    let onPartialResult: @Sendable (STTTranscriptionResult) -> Void
    private let _finalText = OSAllocatedUnfairLock(initialState: "")

    var finalText: String {
        get { _finalText.withLock { $0 } }
        set { _finalText.withLock { $0 = newValue } }
    }

    init(onPartialResult: @Sendable @escaping (STTTranscriptionResult) -> Void) {
        self.onPartialResult = onPartialResult
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/RunAnywhere+STT.ts
/**
 * RunAnywhere+STT.ts
 *
 * Speech-to-Text extension for RunAnywhere SDK.
 * Matches iOS: RunAnywhere+STT.swift
 */

import { EventBus } from '../Events';
import { requireNativeModule, isNativeModuleAvailable } from '../../native';
import type { STTOptions, STTResult } from '../../types';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import type {
  STTOutput,
  STTPartialResult,
  STTStreamCallback,
  STTStreamOptions,
  TranscriptionMetadata,
} from '../../types/STTTypes';

const logger = new SDKLogger('RunAnywhere.STT');

/**
 * Extended native module type for streaming STT methods
 * These methods are optional and may not be implemented in all backends
 */
interface StreamingSTTNativeModule {
  startStreamingSTT?: (language: string) => Promise<boolean>;
  stopStreamingSTT?: () => Promise<boolean>;
  isStreamingSTT?: () => Promise<boolean>;
}

// ============================================================================
// Speech-to-Text (STT) Extension
// ============================================================================

/**
 * Load an STT model
 */
export async function loadSTTModel(
  modelPath: string,
  modelType: string = 'whisper',
  config?: Record<string, unknown>
): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    logger.warning('Native module not available for loadSTTModel');
    return false;
  }
  const native = requireNativeModule();
  return native.loadSTTModel(
    modelPath,
    modelType,
    config ? JSON.stringify(config) : undefined
  );
}

/**
 * Check if an STT model is loaded
 */
export async function isSTTModelLoaded(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule();
  return native.isSTTModelLoaded();
}

/**
 * Unload the current STT model
 */
export async function unloadSTTModel(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule();
  return native.unloadSTTModel();
}

/**
 * Simple voice transcription
 * Matches Swift SDK: RunAnywhere.transcribe(_:)
 *
 * @param audioData Audio data (base64 string or ArrayBuffer)
 * @returns Transcribed text
 */
export async function transcribeSimple(
  audioData: string | ArrayBuffer
): Promise<string> {
  const result = await transcribe(audioData);
  return result.text;
}

/**
 * Transcribe audio data with full options
 * Matches Swift SDK: RunAnywhere.transcribeWithOptions(_:options:)
 */
export async function transcribe(
  audioData: string | ArrayBuffer,
  options?: STTOptions
): Promise<STTResult> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  const native = requireNativeModule();

  let audioBase64: string;
  if (typeof audioData === 'string') {
    audioBase64 = audioData;
  } else {
    const bytes = new Uint8Array(audioData);
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      const byte = bytes[i];
      if (byte !== undefined) {
        binary += String.fromCharCode(byte);
      }
    }
    audioBase64 = btoa(binary);
  }

  const sampleRate = options?.sampleRate ?? 16000;
  const language = options?.language;

  const resultJson = await native.transcribe(audioBase64, sampleRate, language);

  try {
    const result = JSON.parse(resultJson);
    if (result.error) {
      throw new Error(result.error);
    }
    return {
      text: result.text ?? '',
      segments: result.segments ?? [],
      language: result.language,
      confidence: result.confidence ?? 1.0,
      duration: result.duration ?? 0,
      alternatives: result.alternatives ?? [],
    };
  } catch (err) {
    if (err instanceof Error) throw err;
    if (resultJson.includes('error')) {
      const errorMatch = resultJson.match(/"error":\s*"([^"]+)"/);
      throw new Error(errorMatch ? errorMatch[1] : resultJson);
    }
    return {
      text: resultJson,
      segments: [],
      confidence: 1.0,
      duration: 0,
      alternatives: [],
    };
  }
}

/**
 * Transcribe audio buffer (Float32Array)
 * Matches Swift SDK: RunAnywhere.transcribeBuffer(_:language:)
 */
export async function transcribeBuffer(
  samples: Float32Array,
  options?: STTOptions
): Promise<STTOutput> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }

  const startTime = Date.now();
  const native = requireNativeModule();

  // Convert Float32Array to base64
  const bytes = new Uint8Array(samples.buffer);
  let binary = '';
  for (let i = 0; i < bytes.byteLength; i++) {
    const byte = bytes[i];
    if (byte !== undefined) {
      binary += String.fromCharCode(byte);
    }
  }
  const audioBase64 = btoa(binary);

  const sampleRate = options?.sampleRate ?? 16000;
  const language = options?.language ?? 'en';

  const resultJson = await native.transcribe(audioBase64, sampleRate, language);
  const endTime = Date.now();
  const processingTime = (endTime - startTime) / 1000;

  try {
    const result = JSON.parse(resultJson);

    // Estimate audio length from samples
    const audioLength = samples.length / sampleRate;

    const metadata: TranscriptionMetadata = {
      modelId: 'unknown',
      processingTime,
      audioLength,
      realTimeFactor: processingTime / audioLength,
    };

    return {
      text: result.text ?? '',
      confidence: result.confidence ?? 1.0,
      wordTimestamps: result.timestamps,
      detectedLanguage: result.language,
      alternatives: result.alternatives,
      metadata,
    };
  } catch {
    throw new Error(`Transcription failed: ${resultJson}`);
  }
}

/**
 * Transcribe audio with streaming callbacks
 * Matches Swift SDK: RunAnywhere.transcribeStream(audioData:options:onPartialResult:)
 *
 * @param audioData Audio data to transcribe
 * @param options Stream options with callback
 * @returns Final transcription output
 */
export async function transcribeStream(
  audioData: string | ArrayBuffer,
  options: STTStreamOptions
): Promise<STTOutput> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }

  const startTime = Date.now();
  const native = requireNativeModule();

  let audioBase64: string;
  let audioSize: number;

  if (typeof audioData === 'string') {
    audioBase64 = audioData;
    audioSize = atob(audioData).length;
  } else {
    const bytes = new Uint8Array(audioData);
    audioSize = bytes.byteLength;
    let binary = '';
    for (let i = 0; i < bytes.byteLength; i++) {
      const byte = bytes[i];
      if (byte !== undefined) {
        binary += String.fromCharCode(byte);
      }
    }
    audioBase64 = btoa(binary);
  }

  const sampleRate = options?.sampleRate ?? 16000;
  const language = options?.language ?? 'en';

  // Set up event listener for partial results
  let finalText = '';
  let finalConfidence = 1.0;

  if (options.onPartialResult) {
    const unsubscribe = EventBus.onVoice((event) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const evt = event as any;
      if (evt.type === 'sttPartialResult') {
        const partialResult: STTPartialResult = {
          transcript: evt.text ?? '',
          confidence: evt.confidence,
          isFinal: false,
        };
        options.onPartialResult?.(partialResult);
      } else if (evt.type === 'sttCompleted') {
        finalText = evt.text ?? '';
        finalConfidence = evt.confidence ?? 1.0;
        unsubscribe();
      }
    });
  }

  // Transcribe
  const resultJson = await native.transcribe(audioBase64, sampleRate, language);
  const endTime = Date.now();
  const processingTime = (endTime - startTime) / 1000;

  try {
    const result = JSON.parse(resultJson);

    // Estimate audio length
    const bytesPerSample = 2; // 16-bit
    const audioLength = audioSize / (sampleRate * bytesPerSample);

    const metadata: TranscriptionMetadata = {
      modelId: 'unknown',
      processingTime,
      audioLength,
      realTimeFactor: processingTime / audioLength,
    };

    // Emit final partial result
    if (options.onPartialResult) {
      options.onPartialResult({
        transcript: result.text ?? '',
        confidence: result.confidence ?? 1.0,
        isFinal: true,
      });
    }

    return {
      text: result.text ?? finalText,
      confidence: result.confidence ?? finalConfidence,
      wordTimestamps: result.timestamps,
      detectedLanguage: result.language,
      alternatives: result.alternatives,
      metadata,
    };
  } catch {
    throw new Error(`Streaming transcription failed: ${resultJson}`);
  }
}

/**
 * Transcribe audio from a file path
 */
export async function transcribeFile(
  filePath: string,
  options?: STTOptions
): Promise<STTResult> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  const native = requireNativeModule();

  const language = options?.language ?? 'en';
  const resultJson = await native.transcribeFile(filePath, language);

  try {
    const result = JSON.parse(resultJson);
    if (result.error) {
      throw new Error(result.error);
    }
    return {
      text: result.text ?? '',
      segments: result.segments ?? [],
      language: result.language,
      confidence: result.confidence ?? 1.0,
      duration: result.duration ?? 0,
      alternatives: result.alternatives ?? [],
    };
  } catch {
    if (resultJson.includes('error')) {
      const errorMatch = resultJson.match(/"error":\s*"([^"]+)"/);
      throw new Error(errorMatch ? errorMatch[1] : resultJson);
    }
    return {
      text: resultJson,
      segments: [],
      confidence: 1.0,
      duration: 0,
      alternatives: [],
    };
  }
}

// ============================================================================
// Streaming STT (Real-time)
// ============================================================================

/**
 * Start streaming speech-to-text transcription
 * @deprecated Use transcribeStream() for better API parity with Swift SDK
 */
export async function startStreamingSTT(
  language: string = 'en',
  onPartial?: (text: string, confidence: number) => void,
  onFinal?: (text: string, confidence: number) => void,
  onError?: (error: string) => void
): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    logger.warning('Native module not available for startStreamingSTT');
    return false;
  }
  const native = requireNativeModule();

  if (onPartial || onFinal || onError) {
    EventBus.onVoice((event) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const evt = event as any;
      if (evt.type === 'sttPartialResult' && onPartial) {
        onPartial(evt.text || '', evt.confidence || 0);
      } else if (evt.type === 'sttCompleted' && onFinal) {
        onFinal(evt.text || '', evt.confidence || 0);
      } else if (evt.type === 'sttFailed' && onError) {
        onError(evt.error || 'Unknown error');
      }
    });
  }

  const streamingNative = native as unknown as StreamingSTTNativeModule;
  if (!streamingNative.startStreamingSTT) {
    logger.warning('startStreamingSTT not available');
    return false;
  }
  return streamingNative.startStreamingSTT(language);
}

/**
 * Stop streaming speech-to-text transcription
 */
export async function stopStreamingSTT(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule() as unknown as StreamingSTTNativeModule;
  if (!native.stopStreamingSTT) {
    return false;
  }
  return native.stopStreamingSTT();
}

/**
 * Check if streaming STT is currently active
 */
export async function isStreamingSTT(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule() as unknown as StreamingSTTNativeModule;
  if (!native.isStreamingSTT) {
    return false;
  }
  return native.isStreamingSTT();
}
